File: iosApp/iosApp/Modules/CurrencyAdd/CurrencyAddInteractor.swift
// Created by web3d4v on 20/06/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

// MARK: - CurrencyAddInteractorNewToken

struct CurrencyAddInteractorNewCurrency {
    let address: String
    let name: String
    let symbol: String
    let decimals: Int
}

// MARK: - CurrencyAddInteractor

protocol CurrencyAddInteractor: AnyObject {
    func addCurrency(
        _ currency: CurrencyAddInteractorNewCurrency,
        for network: Network,
        onCompletion: @escaping () -> Void
    )
}

// MARK: - DefaultTokenAddInteractor

final class DefaultTokenAddInteractor {
    private let currencyStoreService: CurrencyStoreService

    init(currencyStoreService: CurrencyStoreService) {
        self.currencyStoreService = currencyStoreService
    }
}

extension DefaultTokenAddInteractor: CurrencyAddInteractor {
    
    func addCurrency(
        _ currency: CurrencyAddInteractorNewCurrency,
        for network: Network,
        onCompletion: @escaping () -> Void
    ) {
        let currency = Currency(
            name: currency.name,
            symbol: currency.symbol.lowercased(),
            decimals: KotlinUInt(value: UInt32(currency.decimals)),
            type: .erc20,
            address: currency.address,
            coinGeckoId: nil
        )
        currencyStoreService.add(currency: currency, network: network)
        onCompletion()
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/currencyAdd/CurrencyAddInteractor.kt
package com.sonsofcrypto.web3walletcore.modules.currencyAdd

import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.types.Currency
import com.sonsofcrypto.web3lib.types.Currency.Type.ERC20
import com.sonsofcrypto.web3lib.types.Network

interface CurrencyAddInteractor {
    fun addCurrency(
        contractAddress: String, name: String, symbol: String, decimals: UInt, network: Network
    )
}

class DefaultCurrencyAddInteractor(
    val currencyStoreService: CurrencyStoreService
): CurrencyAddInteractor {

    override fun addCurrency(
        contractAddress: String, name: String, symbol: String, decimals: UInt, network: Network
    ) {
        currencyStoreService.add(
            Currency(name, symbol.lowercase(), decimals, contractAddress, null),
            network
        )
    }
}